warehouse/internal/repo/schema.go

```diff
@@ -192,3 +193,56 @@ func (repo *WHSchema) GetNamespace(ctx context.Context, sourceID, destID string)
 
 	return namespace, nil
 }
+
+func (repo *WHSchema) GetTablesForConnection(ctx context.Context, connections []warehouseutils.Connection) ([]warehouseutils.Connection, error) {
+	if len(connections) == 0 {
+		return nil, fmt.Errorf("no source id and destination id pairs provided")
+	}
+
+	// select max id for each source id and destination id pair
+	innerQueryTemplate := `SELECT max(id) FROM ` + whSchemaTableName + `
+		WHERE
+			(source_id, destination_id) IN (%s)
+		AND
+			namespace <> ''
+		GROUP BY id
+	`
+	sourceIdDestinationIdArr := make([]string, len(connections))
+	for idx, connection := range connections {
+		pair := fmt.Sprintf("'%s','%s'", connection.SourceId, connection.DestinationId)
```

Comment:
Although it is used by an internal endpoint now, I would prefer to write this query in a way to avoid SQL injections. We are not sure how this repository method can be used in the future.

I haven't tried `pq.Array` with nested elements, but it might work.

Another approach could be to compare a concatenated version of them, with single dimension pq.Array

Reply:
+1

Reply:
Can we try something like this?

```GO
func (repo *WHSchema) GetTablesForConnection(ctx context.Context, connections []warehouseutils.SourceIDDestinationID) ([]warehouseutils.FetchTableInfo, error) {
	if len(connections) == 0 {
		return nil, fmt.Errorf("no source id and destination id pairs provided")
	}

	sourceIDDestinationIDPairs := make([]string, len(connections))
	var parameters []interface{}
	for i, conn := range connections {
		sourceIDDestinationIDPairs[i] = fmt.Sprintf("($%d, $%d)", i*2+1, i*2+2)
		parameters = append(parameters, conn.SourceID, conn.DestinationID)
	}

	// select all rows with max id for each source id and destination id pair
	query := `SELECT` + whSchemaTableColumns + `FROM ` + whSchemaTableName + `
		WHERE (source_id, destination_id) IN (` + strings.Join(sourceIDDestinationIDPairs, ", ") + `)`

	rows, err := repo.db.QueryContext(
		ctx,
		query,
		parameters...
	)
	if err != nil {
		return nil, fmt.Errorf("querying schema: %w", err)
	}

	entries, err := repo.parseRows(rows)
	if err != nil {
		return nil, fmt.Errorf("parsing rows: %w", err)
	}

	var tables []warehouseutils.FetchTableInfo
	for _, entry := range entries {
		var allTablesOfConnections []string
		for tableName := range entry.Schema {
			allTablesOfConnections = append(allTablesOfConnections, tableName)
		}
		tables = append(tables, warehouseutils.FetchTableInfo{
			SourceID:      entry.SourceID,
			DestinationID: entry.DestinationID,
			Tables:        allTablesOfConnections,
		})
	}

	return tables, nil
}
```

Reply:
@lvrach  @achettyiitr Thank you for reviewing
I have followed the suggestion from @achettyiitr to avoid SQL injection.